File: Packages/RallyUI/Sources/RallyUI/Components/RallyCard.swift
import SwiftUI
import RallyCore

/// Elevated container with rounded corners and shadow, available in light, dark,
/// or branded style variants.
///
/// Usage:
/// ```swift
/// RallyCard(style: .dark) {
///     Text("Card Content")
/// }
/// ```
public struct RallyCard<Content: View>: View {

    // MARK: - Style

    /// Visual style variants for the card.
    public enum Style {
        /// Light background using off-white.
        case light
        /// Dark background using navy mid.
        case dark
        /// Tinted background derived from the active school theme.
        case branded
    }

    // MARK: - Properties

    @Environment(ThemeEngine.self) private var themeEngine

    private let style: Style
    private let padding: CGFloat
    private let content: () -> Content

    // MARK: - Init

    /// Creates a Rally card.
    /// - Parameters:
    ///   - style: The visual variant (default `.dark`).
    ///   - padding: Inner padding applied to the content (default `RallySpacing.md`).
    ///   - content: The view builder for the card's content.
    public init(
        style: Style = .dark,
        padding: CGFloat = RallySpacing.md,
        @ViewBuilder content: @escaping () -> Content
    ) {
        self.style = style
        self.padding = padding
        self.content = content
    }

    // MARK: - Body

    public var body: some View {
        content()
            .padding(padding)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(
                RoundedRectangle(cornerRadius: RadiusToken.card, style: .continuous)
                    .fill(backgroundColor)
            )
            .shadow(
                color: shadowColor,
                radius: ShadowToken.cardRadius,
                x: ShadowToken.cardX,
                y: ShadowToken.cardY
            )
            .accessibilityElement(children: .contain)
    }

    // MARK: - Appearance Helpers

    private var backgroundColor: Color {
        switch style {
        case .light:
            return RallyColors.offWhite
        case .dark:
            return RallyColors.navyMid
        case .branded:
            return themeEngine.activeTheme.primaryColor.opacity(0.12)
        }
    }

    private var shadowColor: Color {
        switch style {
        case .light:
            return ShadowToken.cardColor
        case .dark:
            return Color.black.opacity(0.12)
        case .branded:
            return themeEngine.activeTheme.primaryColor.opacity(0.1)
        }
    }
}

// MARK: - Preview

#Preview("Card Styles") {
    ScrollView {
        VStack(spacing: 16) {
            RallyCard(style: .dark) {
                VStack(alignment: .leading, spacing: 8) {
                    Text("Dark Card")
                        .font(RallyTypography.cardTitle)
                        .foregroundStyle(.white)
                    Text("Elevated container with navy mid background.")
                        .font(RallyTypography.body)
                        .foregroundStyle(RallyColors.gray)
                }
            }

            RallyCard(style: .light) {
                VStack(alignment: .leading, spacing: 8) {
                    Text("Light Card")
                        .font(RallyTypography.cardTitle)
                        .foregroundStyle(RallyColors.navy)
                    Text("Off-white background for lighter contexts.")
                        .font(RallyTypography.body)
                        .foregroundStyle(RallyColors.gray)
                }
            }

            RallyCard(style: .branded) {
                VStack(alignment: .leading, spacing: 8) {
                    Text("Branded Card")
                        .font(RallyTypography.cardTitle)
                        .foregroundStyle(.white)
                    Text("Tinted with the active school theme color.")
                        .font(RallyTypography.body)
                        .foregroundStyle(RallyColors.gray)
                }
            }
        }
        .padding()
    }
    .background(RallyColors.navy)
    .environment(ThemeEngine())
}

File: Android/ui/src/main/java/com/rally/app/ui/components/RallyCard.kt
package com.rally.app.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import com.rally.app.ui.theme.RallyColors
import com.rally.app.ui.theme.RallyRadius
import com.rally.app.ui.theme.RallyShadow
import com.rally.app.ui.theme.RallySpacing
import com.rally.app.ui.theme.RallyTypography

// ---------------------------------------------------------------------------
// RallyCard
// Elevated container with 16dp rounded corners and shadow.
// Styles: Light (off-white), Dark (navy mid), Branded (school primary tinted).
// ---------------------------------------------------------------------------

/** Visual style variants for the card. */
enum class RallyCardStyle {
    /** Light background using off-white. */
    Light,
    /** Dark background using navy mid. */
    Dark,
    /** Tinted background derived from the active school theme. */
    Branded,
}

/**
 * Rally design-system card.
 *
 * @param modifier       Optional [Modifier].
 * @param style          Visual variant (default [RallyCardStyle.Dark]).
 * @param contentPadding Inner padding applied to the content (default [RallySpacing.md]).
 * @param brandColor     The school primary color, used only when [style] is [RallyCardStyle.Branded].
 * @param content        Composable content slot.
 */
@Composable
fun RallyCard(
    modifier: Modifier = Modifier,
    style: RallyCardStyle = RallyCardStyle.Dark,
    contentPadding: Dp = RallySpacing.md,
    brandColor: Color = RallyColors.Orange,
    content: @Composable () -> Unit,
) {
    val backgroundColor = when (style) {
        RallyCardStyle.Light   -> RallyColors.OffWhite
        RallyCardStyle.Dark    -> RallyColors.NavyMid
        RallyCardStyle.Branded -> brandColor.copy(alpha = 0.12f)
    }
    val shadowElevation = when (style) {
        RallyCardStyle.Light -> RallyShadow.cardElevation
        RallyCardStyle.Dark  -> RallyShadow.cardElevation
        RallyCardStyle.Branded -> RallyShadow.cardElevation
    }

    Surface(
        modifier = modifier.fillMaxWidth(),
        shape = RoundedCornerShape(RallyRadius.card),
        color = backgroundColor,
        shadowElevation = shadowElevation,
    ) {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .padding(contentPadding),
            contentAlignment = Alignment.TopStart,
        ) {
            content()
        }
    }
}

// MARK: - Previews

@Preview(name = "RallyCard - All Styles", showBackground = true, backgroundColor = 0xFF131B2E)
@Composable
private fun RallyCardPreview() {
    Surface(color = RallyColors.Navy) {
        Column(
            modifier = Modifier.padding(RallySpacing.md),
            verticalArrangement = Arrangement.spacedBy(RallySpacing.md),
        ) {
            RallyCard(style = RallyCardStyle.Dark) {
                Column(verticalArrangement = Arrangement.spacedBy(RallySpacing.sm)) {
                    Text(
                        text = "Dark Card",
                        style = RallyTypography.cardTitle,
                        color = Color.White,
                    )
                    Text(
                        text = "Elevated container with navy mid background.",
                        style = RallyTypography.body,
                        color = RallyColors.Gray,
                    )
                }
            }

            RallyCard(style = RallyCardStyle.Light) {
                Column(verticalArrangement = Arrangement.spacedBy(RallySpacing.sm)) {
                    Text(
                        text = "Light Card",
                        style = RallyTypography.cardTitle,
                        color = RallyColors.Navy,
                    )
                    Text(
                        text = "Off-white background for lighter contexts.",
                        style = RallyTypography.body,
                        color = RallyColors.Gray,
                    )
                }
            }

            RallyCard(style = RallyCardStyle.Branded) {
                Column(verticalArrangement = Arrangement.spacedBy(RallySpacing.sm)) {
                    Text(
                        text = "Branded Card",
                        style = RallyTypography.cardTitle,
                        color = Color.White,
                    )
                    Text(
                        text = "Tinted with the active school theme color.",
                        style = RallyTypography.body,
                        color = RallyColors.Gray,
                    )
                }
            }
        }
    }
}
